usr.bin/calendar/calendar.1
.\"	$NetBSD$
.\"
.\" Copyright (c) 1989, 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)calendar.1	8.1 (Berkeley) 6/29/93
.\"
.Dd June 1, 2016
.Dt CALENDAR 1
.Os
.Sh NAME
.Nm calendar
.Nd reminder service
.Sh SYNOPSIS
.Nm
.Op Fl avx
.Op Fl d Ar MMDD[[YY]YY]
.Op Fl f Ar file
.Op Fl l Ar days
.Op Fl w Ar days
.Sh DESCRIPTION
The
.Nm
utility processes text files and displays lines that match certain dates.
.Pp
The following options are available:
.Bl -tag -width Ds
.It Fl a
Process the
.Dq calendar
files of all users and mail the results to them.
This requires super-user privileges.
.It Fl d Ar MMDD[[YY]YY]
Display lines for the given date.
By default, the current date is used.
The year, which may be given in either two or four digit format, is used
only for purposes of determining whether the given date falls on a Friday
in that year (see below).
If the year is not specified, the current year is assumed.
.It Fl f Ar file
Display matching calendar files from the given filename.
By default, the following filenames are checked for:
.Bl -item -compact -offset indent
.It
.Pa ~/calendar
.It
.Pa ~/.calendar
.It
.Pa /etc/calendar
.El
and the first which is found is used.
The filename may be absolute.
If not absolute, it is taken relative to the
directory specified by the
.Ev CALENDAR_DIR
environment variable, if set; otherwise, it is taken relative to the
user's home directory.
Or, if the
.Fl a
flag is given, a non-absolute filename is taken relative to each user's
home directory in turn.
.It Fl l Ar days
Causes the program to
.Dq look ahead
a given number of days (default one) from the specified date and
display their entries as well.
.It Fl w Ar days
Causes the program to add the specified number of days to the
.Dq look ahead
number if and only if the day specified is a Friday.
The default value is two, which causes
.Nm
to print entries through the weekend on Fridays.
.It Fl v
Causes
.Nm
to print version information for itself,
and then exit.
.It Fl x
Causes
.Nm
not to set the
.Ev CPP_RESTRICTED
environment variable.
Passing this flag allows users the (somewhat obscure) option of
including a named pipe via
.Xr cpp 1 Ns No 's
.Li #include
syntax, but opens up the possibility of
.Nm
hanging indefinitely if users do so incorrectly.
For this reason, the
.Fl x
flag should never be used with
.Nm
.Fl a .
.El
.Pp
Lines should begin with a month and day.
They may be entered in almost any format, either numeric or as character
strings.
A single asterisk
.Pq Sq *
matches every month, or every day if a month has been provided.
This means that two asterisks
.Pq Sq **
matches every day of the year, and is thus useful for ToDo tasks.
A day without a month matches that day of every week.
A month without a day matches the first of that month.
Two numbers default to the month followed by the day.
Lines with leading tabs default to the last entered date, allowing
multiple line specifications for a single date.
By convention, dates followed by an asterisk are not fixed, i.e., change
from year to year.
.Pp
The
.Dq calendar
file is preprocessed by
.Xr cpp 1 ,
allowing the inclusion of shared files such as company holidays or
meetings.
If the shared file is not referenced by a full pathname,
.Xr cpp 1
searches in the current (or home) directory first, and then in the
directory
.Pa /usr/share/calendar .
Empty lines and lines protected by the C commenting syntax
.Pq Li /* ... */
are ignored.
.Pp
Some possible calendar entries:
.Bd -literal -offset indent
#include	\*[Lt]calendar.usholiday\*[Gt]
#include	\*[Lt]calendar.birthday\*[Gt]

6/15		... June 15 (if ambiguous, will default to month/day).
Jun. 15		... June 15.
15 June		... June 15.
Thursday	... Every Thursday.
June		... Every June 1st.
15 *		... 15th of every month.
*15		... 15th of every month.
June*		... Every day of June.
**		... Every day
.Ed
.Sh FILES
The following default calendar files are provided:
.Pp
.Bl -tag -width calendar.christian -compact
.It Pa calendar.birthday
Births and deaths of famous (and not-so-famous) people.
.It Pa calendar.christian
Christian holidays.
This calendar should be updated yearly by the local system administrator
so that roving holidays are set correctly for the current year.
.It Pa calendar.computer
Days of special significance to computer people.
.It Pa calendar.history
Everything else, mostly U.S. historical events.
.It Pa calendar.holiday
Other holidays, including the not-well-known, obscure, and
.Em really
obscure.
.It Pa calendar.judaic
Jewish holidays.
This calendar should be updated yearly by the local system administrator
so that roving holidays are set correctly for the current year.
.It Pa calendar.lotr
Important dates in the Lord of the Rings series.
.It Pa calendar.music
Musical events, births, and deaths.
Strongly oriented toward rock 'n' roll.
.It Pa calendar.netbsd
Important dates in the history of the
.Nx
project.
Mostly releases and port additions.
.It Pa calendar.usholiday
U.S. holidays.
This calendar should be updated yearly by the local system administrator
so that roving holidays are set correctly for the current year.
.El
.Sh COMPATIBILITY
The
.Nm
program previously selected lines which had the correct date anywhere
in the line.
This is no longer true, the date is only recognized when it occurs
first on the line.
.Pp
In
.Nx 3.0 ,
the
.Nm
command was modified to search the user's home directory instead of the current
directory by default.
Users desiring the historical behavior should set the
.Ev CALENDAR_DIR
environment variable to
.Pa \&. ,
or use the
.Fl f
flag.
.Sh SEE ALSO
.Xr at 1 ,
.Xr cpp 1 ,
.Xr cron 8
.Sh HISTORY
A
.Nm
command appeared in
.At v7 .
.Sh BUGS
.Nm
doesn't handle events that move around from year to year, i.e.,
.Dq the last Monday in April .
.Pp
The
.Fl a
option ignores the user's
.Ev CALENDAR_DIR
environment variable.


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision$>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 7.0 to NetBSD 8.0:
	libc: Update to tzcode2014f. [christos 20140815]
	gmake:	Update to gmake-3.81 in external/gpl2/gmake and remove 
		gnu/dist/gmake (3.80) [christos 20140818]
	module(7): Create MODULAR_DEFAULT_AUTOLOAD config flag to control
		whether module autoloading is on by default.
		[jnemeth 20140824]
	kernel: Option SYMTAB_SPACE has been replaced with COPY_SYMTAB.
		The related changes to src/sys/conf require a rebuild of
		config(1). [joerg 20140824]
	zoneinfo: Import tzdata2014g. [apb 20140831]
	wm(4): Add internal SERDES support for 82575 and newer controllers.
		[msaitoh 20140903]
	or1k: Add new port.  [matt 20140903]
	awindma(4): Add support for Allwinner A10/A20 DMA controller.
		[jmcneill 20140906]
	awinac(4): Add support for Allwinner A10/A20 audio codec.
		[jmcneill 20140906]
	awinrtc(4): Add support for Allwinner A10/A20 real-time clock.
		[jmcneill 20140907]
	awinmmc(4): Use DMA for MMC transfers. [jmcneill 20140908]
	axp20x(4): Add AXP20x PMU driver. [jmcneill 20140909]
	rtsol(8): Removed in favour of dhcpcd. [roy 20140911]
	rtsold(8): Removed in favour of dhcpcd. [roy 20140911]
	dhcpcd(8): Import dhcpcd-6.4.5. [roy 20140918]
	arm: Add support for i.MX6 SoC. [ryo 20140925]
	gpt(8): Completed overhaul, including adding follwing subcommands:
		resize, set, unset, backup, restore, and resizedisk.
		[jnemeth 20140926]
	dhcpcd(8): Import dhcpcd-6.4.7. [roy 20140927]
	mpt(4): added bio(4) support. [jmcneill 20140927]
	gpt(8): added type subcommand to modify partition type.
		[jnemeth 20140927]
	zoneinfo: Import tzdata2014h. [apb 20141004]
	dhcpcd(8): Import dhcpcd-6.5.0. [roy 20141006]
	libc: Update to tzcode2014h. [christos 20141007]
	file(1): Upgraded to 5.20. [christos 20141010]
	arm: Add support for Allwinner A31 SoC. [jmcneill 20141010]
	awiniic(4): Add support for Allwinner A31 SoC. [jmcneill 20141012]
	awinp2wi(4): Add support for Allwinner A31 Push-Pull Two Wire
		Interface. [jmcneill 20141012]
	axp22x(4): Add AXP22x PMU driver. [jmcneill 20141012]
	awinrtc(4): Add support for Allwinner A31 real-time clock.
		[jmcneill 20141012]
	awindma(4): Add support for Allwinner A31 DMA controller.
		[jmcneill 20141013]
	awinac(4): Add support for Allwinner A31 audio codec.
		[jmcneill 20141013]
	motg(4): Add support for Allwinner A10/A20 SoC. [jmcneill 20141015]
	wpa: Import wpa_supplicant and hostapd 2.3. [christos 20141016]
	dhcpcd(8): Import dhcpcd-6.5.1. [roy 20141018]
	OpenSSH: Imported 6.7. [christos 20141018]
	resolvconf(8): Import openresolv-3.6.0 [roy 20141020]
	gcc: Support -fsanitize=address [christos 20141022]
	zoneinfo: Import tzdata2014i. [apb 20141023]
	libc: Update to tzcode2014i. [christos 20141023]
	openpam(3): update to 20140912 (ourouparia) [christos 20141024]
	pppd(8): updated to version 2.4.7. [christos 20141025]
	acpi(4): Updated ACPICA to 20140926. [christos 20141025]
	dhcpcd(8): Import dhcpcd-6.6.0 [roy 20141029]
	resolvconf(8): Import openresolv-3.6.1 [roy 20141029]
	rtsx(4): Add support for Realtek RTS5227/RTL8402/RTL8411/RTL8411B.

usr.bin/calendar/calendar.c
/*	$NetBSD$	*/

/*
 * Copyright (c) 1989, 1993, 1994
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1989, 1993\
 The Regents of the University of California.  All rights reserved.");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)calendar.c	8.4 (Berkeley) 1/7/95";
#endif
__RCSID("$NetBSD$");
#endif /* not lint */

#include <sys/param.h>
#include <sys/ioctl.h>
#include <sys/time.h>
#include <sys/stat.h>
#include <sys/uio.h>
#include <sys/wait.h>

#include <assert.h>
#include <ctype.h>
#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <pwd.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <tzfile.h>
#include <unistd.h>

#include "pathnames.h"

#define CALENDAR_VERSION	"calendar-20160601"

	/* flags used by calendar file parser */
#define	F_ISMONTH	0x01
#define	F_ISDAY		0x02
#define	F_ISDOW		0x04
#define	F_WILDMONTH	0x10
#define	F_WILDDAY	0x20

static unsigned short lookahead = 1;
static unsigned short weekend = 2;
static char *fname = NULL;
static char *datestr = NULL;
static const char *defaultnames[] = {"calendar", ".calendar", _PATH_SYSTEM_CALENDAR, NULL};
static struct passwd *pw;
static char path[MAXPATHLEN + 1];
static bool doall = false;
static bool cpp_restricted = false;

/* 1-based month, 0-based days, cumulative */
static const int daytab[][14] = {
	{ 0, -1, 30, 58, 89, 119, 150, 180, 211, 242, 272, 303, 333, 364 },
	{ 0, -1, 30, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334, 365 },
};
static struct tm *tp;
static const int *cumdays;
static int offset, yrdays;
static char dayname[10];

static struct iovec header[] = {
	{ __UNCONST("From: "), 6 },
	{ NULL, 0 },
	{ __UNCONST(" (Reminder Service)\nTo: "), 24 },
	{ NULL, 0 },
	{ __UNCONST("\nSubject: "), 10 },
	{ NULL, 0 },
	{ __UNCONST("'s Calendar\nPrecedence: bulk\n\n"),  30 },
};

static const char *days[] = {
	"sun", "mon", "tue", "wed", "thu", "fri", "sat", NULL,
};

static const char *months[] = {
	"jan", "feb", "mar", "apr", "may", "jun",
	"jul", "aug", "sep", "oct", "nov", "dec", NULL,
};

static void	 atodays(int, char *, unsigned short *);
static void	 cal(void);
static void	 closecal(FILE *);
static void	 changeuser(void);
static int	 getday(char *);
static int	 getfield(char *, char **, int *);
static void	 getmmdd(struct tm *, char *);
static int	 getmonth(char *);
static bool	 isnow(char *);
static FILE	*opencal(FILE **);
static int	 tryopen(const char *, int);
static void	 settime(void);
static void	 usage(void) __dead;

int
main(int argc, char **argv)
{
	int ch;
	const char *caldir;

	(void)setprogname(argv[0]);	/* for portability */

	while ((ch = getopt(argc, argv, "-ad:f:l:vw:x")) != -1) {
		switch (ch) {
		case '-':		/* backward contemptible */
		case 'a':
			if (getuid()) {
				errno = EPERM;
				err(EXIT_FAILURE, NULL);
			}
			doall = true;
			break;
		case 'd':
			datestr = optarg;
			break;
		case 'f':
			fname = optarg;
			break;
		case 'l':
			atodays(ch, optarg, &lookahead);
			break;
		case 'v':
			printf("%s\n", CALENDAR_VERSION);
			return 0;
		case 'w':
			atodays(ch, optarg, &weekend);
			break;
		case 'x':
			cpp_restricted = true;
			break;
		case '?':
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	if (argc)
		usage();

	settime();
	if (doall) {
		/*
		 * XXX - This ignores the user's CALENDAR_DIR variable.
		 *       Run under user's login shell?
		 */
		if (setgroups(0, NULL) == -1) {
			err(EXIT_FAILURE, "setgroups");
		}
		while ((pw = getpwent()) != NULL) {
			if (setegid(pw->pw_gid) == -1) {
				warn("%s: setegid", pw->pw_name);
				continue;
			}
			if (seteuid(pw->pw_uid) == -1) {
				warn("%s: seteuid", pw->pw_name);
				continue;
			}
			if (chdir(pw->pw_dir) != -1) {
				cal();
			}
			if (seteuid(0) == -1) {
				warn("%s: seteuid back to 0", pw->pw_name);
			}
		}
	} else if ((caldir = getenv("CALENDAR_DIR")) != NULL) {
		if (chdir(caldir) != -1)
			cal();
	} else if ((pw = getpwuid(geteuid())) != NULL) {
		if (chdir(pw->pw_dir) != -1)
			cal();
	}
	return 0;
}

static void
cal(void)
{
	bool printing;
	FILE *fp, *in = NULL;
	char *line;

	if ((fp = opencal(&in)) == NULL || in == NULL)
		return;
	printing = false;
	while ((line = fparseln(in,
		    NULL, NULL, NULL, FPARSELN_UNESCCOMM)) != NULL) {
		if (line[0] == '\0')
			continue;
		if (line[0] != '\t')
			printing = isnow(line);
		if (printing)
			(void)fprintf(fp, "%s\n", line);
		free(line);

	}
	closecal(fp);
}

static void
settime(void)
{
	time_t now;

	(void)time(&now);
	tp = localtime(&now);
	if (datestr)
		getmmdd(tp, datestr);

	if (isleap(tp->tm_year + TM_YEAR_BASE)) {
		yrdays = DAYSPERLYEAR;
		cumdays = daytab[1];
	} else {
		yrdays = DAYSPERNYEAR;
		cumdays = daytab[0];
	}
	/* Friday displays Monday's events */
	offset = tp->tm_wday == 5 ? lookahead + weekend : lookahead;
	header[5].iov_base = dayname;
	header[5].iov_len = strftime(dayname, sizeof(dayname), "%A", tp);
}

/*
 * Possible date formats include any combination of:
 *	3-charmonth			(January, Jan, Jan)
 *	3-charweekday			(Friday, Monday, mon.)
 *	numeric month or day		(1, 2, 04)
 *
 * Any character may separate them, or they may not be separated.  Any line,
 * following a line that is matched, that starts with "whitespace", is shown
 * along with the matched line.
 */
static bool
isnow(char *endp)
{
	int day;
	int flags;
	int month;
	int v1;
	int v2;

	flags = 0;

	/* didn't recognize anything, skip it */
	if (!(v1 = getfield(endp, &endp, &flags)))
		return false;

	if ((flags & (F_ISDAY|F_ISDOW)) || v1 > 12) {
		/* found a day */
		day = v1;
		/* if no recognizable month, assume wildcard ('*') month */
		if ((month = getfield(endp, &endp, &flags)) == 0) {
			flags |= F_ISMONTH | F_WILDMONTH;
			month = tp->tm_mon + 1;
		}
	} else if (flags & F_ISMONTH) {
		month = v1;
		/* if no recognizable day, assume the first */
		if ((day = getfield(endp, &endp, &flags)) == 0)
			day = 1;
	} else {
		v2 = getfield(endp, &endp, &flags);
		if (flags & F_ISMONTH) {
			day = v1;
			month = v2;
		} else {
			/* F_ISDAY set, v2 > 12, or no way to tell */
			month = v1;
			/* if no recognizable day, assume the first */
			day = v2 ? v2 : 1;
		}
	}
	/* if month is out of range, treat it as '*' */
	if (month < 1 || month > 12) {
		flags |= F_ISMONTH | F_WILDMONTH;
		month = tp->tm_mon + 1;
	}

	if (flags & F_WILDMONTH && flags & F_WILDDAY)
		return true;

	if (flags & F_WILDMONTH && flags & F_ISDAY && day == tp->tm_mday)
		return true;

	if (flags & F_WILDMONTH && flags & F_ISDOW && day == tp->tm_wday + 1)
		return true;

	if (flags & F_ISMONTH && flags & F_WILDDAY && month == tp->tm_mon + 1)
		return true;

	if (flags & F_ISMONTH && flags & F_ISDOW && month == tp->tm_mon + 1 &&
	    day == tp->tm_wday + 1)
		return true;

	if (flags & F_ISDOW)
		day = tp->tm_mday + (((day - 1) - tp->tm_wday + 7) % 7);
	day = cumdays[month] + day;

	/* if today or today + offset days */
	if (day >= tp->tm_yday && day <= tp->tm_yday + offset)
		return true;

	/* if number of days left in this year + days to event in next year */
	if (yrdays - tp->tm_yday + day <= offset)
		return true;

	return false;
}

static int
getfield(char *p, char **endp, int *flags)
{
	int val;
	char *start;
	char savech;

/*
 * note this macro has an arg that isn't used ... it is retained
 * (it is believed) to make the macro call look more "natural"
 * and suggest at the call site what is happening.
 */
#define FLDCHAR(a) (*p != '\0' && !isdigit((unsigned char)*p) && \
    !isalpha((unsigned char)*p) && *p != '*')

	val = 0;
	for (/*EMPTY*/; FLDCHAR(*p); ++p)
		continue;
	if (*p == '*') {			/* `*' is current month */
		if (!(*flags & F_ISMONTH)) {
			*flags |= F_ISMONTH | F_WILDMONTH;
			*endp = p + 1;
			return tp->tm_mon + 1;
		} else {
			*flags |= F_ISDAY | F_WILDDAY;
			*endp = p + 1;
			return 1;
		}
	}
	if (isdigit((unsigned char)*p)) {
		val = (int)strtol(p, &p, 10);	/* if 0, it's failure */
		for (/*EMPTY*/; FLDCHAR(*p); ++p)
			continue;
		*endp = p;
		return val;
	}
	for (start = p; *p != '\0' && isalpha((unsigned char)*p); p++)
		continue;

	savech = *p;
	if (p != start) {
		*p = '\0';
		if ((val = getmonth(start)) != 0)
			*flags |= F_ISMONTH;
		else if ((val = getday(start)) != 0)
			*flags |= F_ISDOW;
		else {
			*p = savech;
			*endp = start;
			return 0;
		}
	}
	for (*p = savech; FLDCHAR(*p); ++p)
		continue;
	*endp = p;
	return val;
}

static FILE *
opencal(FILE **in)
{
	int fd = -1;
	int pdes[2];

	/* open up calendar file as stdin */
	if (fname == NULL) {
		for (const char **name = defaultnames; *name != NULL; name++) {
			if ((fd = tryopen(*name, O_RDONLY)) == -1)
				continue;
			else
				break;
		}
		if (fd == -1) {
			if (doall)
				return NULL;
			err(EXIT_FAILURE, "Cannot open calendar file");
		}
	} else if ((fd = tryopen(fname, O_RDONLY)) == -1) {
		if (doall)
			return NULL;
		err(EXIT_FAILURE, "Cannot open `%s'", fname);
	}

	if (pipe(pdes) == -1) {
		warn("Cannot open pipe");
		return NULL;
	}

	switch (fork()) {
	case -1:
		/* error */
		(void)close(pdes[0]);
		(void)close(pdes[1]);
		return NULL;
	case 0:
		/* child */
		/* set stdin to calendar file */
		if (fd != STDIN_FILENO) {
			(void)dup2(fd, STDIN_FILENO);
			(void)close(fd);
		}
		/* set stdout to pipe input */
		if (pdes[1] != STDOUT_FILENO) {
			(void)dup2(pdes[1], STDOUT_FILENO);
			(void)close(pdes[1]);
		}
		(void)close(pdes[0]);
		if (doall) {
			/* become the user properly */
			changeuser();
		}
		/* tell CPP to only open regular files */
		if(!cpp_restricted && setenv("CPP_RESTRICTED", "", 1) == -1)
			err(EXIT_FAILURE, "Cannot restrict cpp");
		cpp_restricted = true;

		(void)execl(_PATH_CPP, "cpp", "-traditional", "-P", "-I.",
		    "-I" _PATH_CALENDARS, NULL);
		err(EXIT_FAILURE, "Cannot exec `%s'", _PATH_CPP);
		/*NOTREACHED*/
	default:
		/* parent -- fdopen *in to pipe output */
		*in = fdopen(pdes[0], "r");
		(void)close(pdes[1]);

		/* close calendar file */
		close(fd);

		/* not reading all calendar files, just set output to stdout */
		if (!doall)
			return stdout;

		/*
		 * Set output to a temporary file, so if no output
		 * don't send mail.
		 */
		(void)snprintf(path, sizeof(path), "%s/_calXXXXXX", _PATH_TMP);
		if ((fd = mkstemp(path)) == -1) {
			warn("Cannot create temporary file");
			return NULL;
		}
		return fdopen(fd, "w+");
	}
	/*NOTREACHED*/
}

static int
tryopen(const char *pathname, int flags)
{
	int fd, serrno, zero;
	struct stat st;

	/*
	 * XXX: cpp_restricted has inverted sense; it is false by default,
	 * and -x sets it to true. CPP_RESTRICTED is set in the environment
	 * if cpp_restricted is false... go figure. This should be fixed
	 * later.
	 */
	if (doall && cpp_restricted == false) {
		/*
		 * We are running with the user's euid, so they can't
		 * cause any mayhem (e.g. opening rewinding tape
		 * devices) that they couldn't do easily enough on
		 * their own. All we really need to worry about is opens
		 * that hang, because that would DoS the calendar run.
		 */
		fd = open(pathname, flags | O_NONBLOCK);
		if (fd == -1) {
			return -1;
		}
		if (fstat(fd, &st) == -1) {
			serrno = errno;
			close(fd);
			errno = serrno;
			return -1;
		}
		if (S_ISCHR(st.st_mode) ||
		    S_ISBLK(st.st_mode) ||
		    S_ISFIFO(st.st_mode)) {
			close(fd);

			/* Call shenanigans in the daily output */
			errno = EPERM;
			warn("%s: %s", pw->pw_name, pathname);

			errno = EPERM;
			return -1;
		}
		if (S_ISDIR(st.st_mode)) {
			/* Don't warn about this */
			close(fd);
			errno = EISDIR;
			return -1;
		}
		if (!S_ISREG(st.st_mode)) {
			/* There shouldn't be other cases to go here */
			close(fd);
			errno = EINVAL;
			return -1;
		}
		zero = 0;
		if (ioctl(fd, FIONBIO, &zero) == -1) {
			serrno = errno;
			warn("%s: %s: FIONBIO", pw->pw_name, pathname);
			close(fd);
			errno = serrno;
			return -1;
		}
		return fd;
	} else {
		return open(pathname, flags);
	}
}

static void
closecal(FILE *fp)
{
	struct stat sbuf;
	ssize_t nread;
	int pdes[2];
	int status;
	char buf[1024];

	if (!doall)
		return;

	(void)rewind(fp);
	if (fstat(fileno(fp), &sbuf) == -1 || sbuf.st_size == 0)
		goto done;
	if (pipe(pdes) == -1)
		goto done;

	switch (fork()) {
	case -1:
		/* error */
		(void)close(pdes[0]);
		(void)close(pdes[1]);
		break;
	case 0:
		/* child -- set stdin to pipe output */
		if (pdes[0] != STDIN_FILENO) {
			(void)dup2(pdes[0], STDIN_FILENO);
			(void)close(pdes[0]);
		}
		(void)close(pdes[1]);
		if (doall) {
			/* become the user properly */
			changeuser();
		}
		(void)execl(_PATH_SENDMAIL, "sendmail", "-i", "-t", "-F",
		    "\"Reminder Service\"", "-f", "root", NULL);
		err(EXIT_FAILURE, "Cannot exec `%s'", _PATH_SENDMAIL);
		/*NOTREACHED*/
	default:
		/* parent -- write to pipe input */
		(void)close(pdes[0]);

		header[1].iov_base = header[3].iov_base = (void *)pw->pw_name;
		header[1].iov_len = header[3].iov_len = strlen(pw->pw_name);
		(void)writev(pdes[1], header, 7);
		while ((nread = read(fileno(fp), buf, sizeof(buf))) > 0)
			(void)write(pdes[1], buf, (size_t)nread);
		(void)close(pdes[1]);
		break;
	}

done:	(void)fclose(fp);
	(void)unlink(path);
	while (wait(&status) != -1)
		continue;
}

static void
changeuser(void)
{
	uid_t uid;
	gid_t gid;

	uid = geteuid();
	gid = getegid();
	assert(uid == pw->pw_uid);
	assert(gid == pw->pw_gid);

	if (seteuid(0) == -1) {
		err(EXIT_FAILURE, "%s: changing user: cannot reassert uid 0",
		    pw->pw_name);
	}
	if (setgid(gid) == -1) {
		err(EXIT_FAILURE, "%s: cannot assume gid %d",
		    pw->pw_name, (int)gid);
	}
	if (initgroups(pw->pw_name, gid) == -1) {
		err(EXIT_FAILURE, "%s: cannot initgroups", pw->pw_name);
	}
	if (setuid(uid) == -1) {
		err(EXIT_FAILURE, "%s: cannot assume uid %d",
		    pw->pw_name, (int)uid);
	}
}

static int
getmonth(char *s)
{
	const char **p;

	for (p = months; *p; ++p)
		if (strncasecmp(s, *p, 3) == 0)
			return (int)(p - months) + 1;
	return 0;
}

static int
getday(char *s)
{
	const char **p;

	for (p = days; *p; ++p)
		if (strncasecmp(s, *p, 3) == 0)
			return (int)(p - days) + 1;
	return 0;
}

static void
atodays(int ch, char *arg, unsigned short *rvp)
{
	int u;

	u = atoi(arg);
	if (u < 0 || u > 366)
		warnx("-%c %d out of range 0-366, ignored.", ch, u);
	else
		*rvp = u;
}

#define todigit(x) ((x) - '0')
#define ATOI2(x) (todigit((x)[0]) * 10 + todigit((x)[1]))
#define ISDIG2(x) (isdigit((unsigned char)(x)[0]) && isdigit((unsigned char)(x)[1]))

static void
getmmdd(struct tm *ptm, char *ds)
{
	bool ok = false;
	struct tm ttm;

	ttm = *ptm;
	ttm.tm_isdst = -1;

	if (ISDIG2(ds)) {
		ttm.tm_mon = ATOI2(ds) - 1;
		ds += 2;
	}
	if (ISDIG2(ds)) {
		ttm.tm_mday = ATOI2(ds);
		ds += 2;
		ok = true;
	}
	if (ok) {
		if (ISDIG2(ds) && ISDIG2(ds + 2)) {
			ttm.tm_year = ATOI2(ds) * 100 - TM_YEAR_BASE;
			ds += 2;
			ttm.tm_year += ATOI2(ds);
		} else if (ISDIG2(ds)) {
			ttm.tm_year = ATOI2(ds);
			if (ttm.tm_year < 69)
				ttm.tm_year += 2000 - TM_YEAR_BASE;
			else
				ttm.tm_year += 1900 - TM_YEAR_BASE;
		}
	}
	if (ok && mktime(&ttm) == -1)
		ok = false;

	if (ok)
		*ptm = ttm;
	else {
		warnx("Can't convert `%s' to date, ignored.", ds);
		usage();
	}
}

__dead
static void
usage(void)
{
	(void)fprintf(stderr, "usage: %s [-ax] [-d MMDD[[YY]YY]"
	    " [-f fname] [-l days] [-w days]\n", getprogname());
	exit(1);
}
